prov/util/src/util_mem_hooks.c

```diff
@@ -125,246 +154,209 @@ static struct ofi_intercept intercepts[] = {
 				.our_func = ofi_intercept_shmdt},
 	[OFI_INTERCEPT_BRK] = { .symbol = "brk",
 				.our_func = ofi_intercept_brk},
-	[OFI_INTERCEPT_SBRK] = { .symbol = "sbrk",
-				.our_func = ofi_intercept_sbrk},
-};
-
-struct ofi_mem_calls {
-	void *(*dlopen) (const char *, int);
-	void *(*mmap)(void *, size_t, int, int, int, off_t);
-	int (*munmap)(void *, size_t);
-	void *(*mremap)(void *old_address, size_t old_size,
-			size_t new_size, int flags, ... /* void *new_address */ );
-	int (*madvise)(void *addr, size_t length, int advice);
-	void *(*shmat)(int shmid, const void *shmaddr, int shmflg);
-	int (*shmdt)(const void *shmaddr);
-	int (*brk)(const void *brkaddr);
-	void *(*sbrk)(intptr_t);
 };
 
-static struct ofi_mem_calls real_calls;
-
+#ifdef HAVE___CURBRK
+extern void *__curbrk; /* in libc */
+#endif
 
-static const ElfW(Phdr) *
-ofi_get_phdr_dynamic(const ElfW(Phdr) *phdr, uint16_t phnum, int phent)
+#if HAVE___CLEAR_CACHE
+/*
+ * Used on ARM64 platforms, see https://github.com/open-mpi/ompi/issues/5631
+ */
+static inline void ofi_clear_instruction_cache(uintptr_t address, size_t data_size)
 {
-	uint16_t i;
-
-	for (i = 0 ; i < phnum; i++) {
-		if (phdr->p_type == PT_DYNAMIC)
-			return phdr;
-		phdr = (ElfW(Phdr)*) ((intptr_t) phdr + phent);
-	}
+	/* do not allow global declaration of compiler intrinsic */
+	void __clear_cache(void* beg, void* end);
 
-	return NULL;
+	__clear_cache ((void *) address, (void *) (address + data_size));
 }
-
-static void *ofi_get_dynentry(ElfW(Addr) base, const ElfW(Phdr) *pdyn,
-			      ElfW(Sxword) type)
+#else
+static inline void ofi_clear_instruction_cache(uintptr_t address, size_t data_size)
 {
-	ElfW(Dyn) *dyn;
-
-	for (dyn = (ElfW(Dyn)*) (base + pdyn->p_vaddr); dyn->d_tag; ++dyn) {
-		if (dyn->d_tag == type)
-			return (void *) (uintptr_t) dyn->d_un.d_val;
+	size_t i;
+	size_t offset_jump = 16;
+#if defined(__aarch64__)
+	offset_jump = 32;
+#endif
+	/* align the address */
+	address &= ~(offset_jump - 1);
+
+	for (i = 0 ; i < data_size ; i += offset_jump) {
+#if (defined(__x86_64__) || defined(__amd64__))
+		__asm__ volatile("mfence;clflush %0;mfence"::
+				 "m" (*((char*) address + i)));
+#elif defined(__aarch64__)
```

Comment:
No fence needed (I am not as familiar with ARM ASM as I might be)?

Reply:
I'm too familiar with ARM assembly myself and was going of the discussion and fix that happened in Open MPI here:
https://github.com/open-mpi/ompi/issues/5631